PPO/PPO.py

```diff
@@ -67,3 +87,85 @@ def learn(self, num_learn):
       self.optimiser.step()
     
     self.model_old.load_state_dict(self.model.state_dict())
+
+class PPOPixel(PPOBase):
+  def __init__(self, config):
+    super(PPOPixel, self).__init__(config)
+    self.config = config
+  
+  def state_shaper(self, state):
+    state = np.array(state).transpose((2, 0, 1))
+    state = torch.FloatTensor(state)
+    state = state.unsqueeze(0)
+    state = state.float() / 256
+
+    return state
+  
+  def add_to_mem(self, state, action, reward, log_prob, done):
+    state = self.state_shaper(state)
+    self.mem.add(state, action, reward, log_prob, done)
+  
+  def act(self, x):
+    x = self.state_shaper(x).to(self.config.device)
+    return self.model_old.act(x)
+
+  def learn(self, num_learn, last_value, next_done):
+    # Calculate discounted rewards
+    bootstrap_length = self.config.update_every
+    discounted_returns = torch.zeros(bootstrap_length)
+    for t in reversed(range(bootstrap_length)):
+      # If first loop
+      if t == bootstrap_length - 1:
+        nextnonterminal = 1.0 - next_done
+        next_return = last_value
+      else:
+        nextnonterminal = 1.0 - self.mem.dones[t+1]
+        next_return = discounted_returns[t+1]
+      discounted_returns[t] = self.mem.rewards[t] + self.gamma * nextnonterminal * next_return
+
+    # normalise rewards
+    discounted_returns = torch.FloatTensor(discounted_returns).to(self.device)
+    discounted_returns = (discounted_returns - discounted_returns.mean()) / (discounted_returns.std() + 1e-8)
+    
+    # Fetch collected state, action and log_probs from memory & reshape for nn
+    prev_states = torch.stack(self.mem.states).reshape((-1,)+(4, 84, 84)).to(self.device).detach()
```

Comment:
Currently hardcoded shape (e.g. won't work with procgen)

Reply:
yeh definitely, the env that I used right now comes out as something weird and I couldn't use it. We can amend this when we do the multi env.